### 1. Импортируйте файл bike.csv и создайте на его основе pandas dataframe (pandas.core.frame.DataFrame). 

In [1]:
import pandas as pd
import numpy as np
from os import getcwd

df = pd.read_csv(fr'{getcwd()}\data\bike.csv', sep=',')

In [5]:
df

datetime  season  holiday  workingday  weather   temp  \
0     2011-01-01 00:00:00       1        0           0      1.0   9.84   
1     2011-01-01 01:00:00       1        0           0      1.0   9.02   
2     2011-01-01 02:00:00       1        0           0      1.0   9.02   
3     2011-01-01 03:00:00       1        0           0      1.0   9.84   
4     2011-01-01 04:00:00       1        0           0      1.0   9.84   
...                   ...     ...      ...         ...      ...    ...   
7800  2012-06-05 19:00:00       2        0           1      1.0  23.78   
7801  2012-06-05 20:00:00       2        0           1      1.0  22.96   
7802  2012-06-05 21:00:00       2        0           1      1.0  22.14   
7803  2012-06-05 22:00:00       2        0           1      1.0  22.96   
7804  2012-06-05 23:00:00       2        0           1      1.0  22.14   

       atemp  humidity  windspeed  casual  registered  count  
0     14.395      81.0     0.0000       3          13     16  
1     13.635      80.0     0.0000       8          32     40  
2     13.635      80.0     0.0000       5          27     32  
3     14.395      75.0     0.0000       3          10     13  
4     14.395      75.0     0.0000       0           1      1  
...      ...       ...        ...     ...         ...    ...  
7800  27.275      46.0    16.9979      68         445    513  
7801  26.515      49.0    15.0013      44         371    415  
7802  25.760      56.0    11.0014      30         253    283  
7803  26.515      52.0     7.0015      22         171    193  
7804  25.760      56.0     6.0032      20          93    113  

[7805 rows x 12 columns]

### 2. Определите количество строк и столбцов в датафрейме. 

In [6]:
df.shape

(7805, 12)

### 3. Выведите первые 6 и последние 7 строк датафрейма. 

Первые 6

In [8]:
df.head(6)

datetime  season  holiday  workingday  weather  temp   atemp  \
0  2011-01-01 00:00:00       1        0           0      1.0  9.84  14.395   
1  2011-01-01 01:00:00       1        0           0      1.0  9.02  13.635   
2  2011-01-01 02:00:00       1        0           0      1.0  9.02  13.635   
3  2011-01-01 03:00:00       1        0           0      1.0  9.84  14.395   
4  2011-01-01 04:00:00       1        0           0      1.0  9.84  14.395   
5  2011-01-01 05:00:00       1        0           0      2.0  9.84  12.880   

   humidity  windspeed  casual  registered  count  
0      81.0     0.0000       3          13     16  
1      80.0     0.0000       8          32     40  
2      80.0     0.0000       5          27     32  
3      75.0     0.0000       3          10     13  
4      75.0     0.0000       0           1      1  
5      75.0     6.0032       0           1      1

Последние 7

In [7]:
df.tail(7)

datetime  season  holiday  workingday  weather   temp  \
7798  2012-06-05 17:00:00       2        0           1      1.0  24.60   
7799  2012-06-05 18:00:00       2        0           1      1.0  24.60   
7800  2012-06-05 19:00:00       2        0           1      1.0  23.78   
7801  2012-06-05 20:00:00       2        0           1      1.0  22.96   
7802  2012-06-05 21:00:00       2        0           1      1.0  22.14   
7803  2012-06-05 22:00:00       2        0           1      1.0  22.96   
7804  2012-06-05 23:00:00       2        0           1      1.0  22.14   

       atemp  humidity  windspeed  casual  registered  count  
7798  31.060      43.0     0.0000      86         764    850  
7799  31.060      43.0    12.9980     111         679    790  
7800  27.275      46.0    16.9979      68         445    513  
7801  26.515      49.0    15.0013      44         371    415  
7802  25.760      56.0    11.0014      30         253    283  
7803  26.515      52.0     7.0015      22         171    193  
7804  25.760      56.0     6.0032      20          93    113

### 4. Выведите список всех столбцов. 

In [9]:
df.columns

Index(['datetime', 'season', 'holiday', 'workingday', 'weather', 'temp',
       'atemp', 'humidity', 'windspeed', 'casual', 'registered', 'count'],
      dtype='object')

### 5. Выведите транспонированный датафрейм (первые 3 элемента). 

In [10]:
df.head(3).T

0                    1                    2
datetime    2011-01-01 00:00:00  2011-01-01 01:00:00  2011-01-01 02:00:00
season                        1                    1                    1
holiday                       0                    0                    0
workingday                    0                    0                    0
weather                     1.0                  1.0                  1.0
temp                       9.84                 9.02                 9.02
atemp                    14.395               13.635               13.635
humidity                   81.0                 80.0                 80.0
windspeed                   0.0                  0.0                  0.0
casual                        3                    8                    5
registered                   13                   32                   27
count                        16                   40                   32

### 6. Выведите информацию обо всех данных. Есть ли пропуски и если есть, то сколько? Какого типа данные представлены в датафрейме? 

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7805 entries, 0 to 7804
Data columns (total 12 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   datetime    7805 non-null   object 
 1   season      7805 non-null   int64  
 2   holiday     7805 non-null   int64  
 3   workingday  7805 non-null   int64  
 4   weather     7803 non-null   float64
 5   temp        7803 non-null   float64
 6   atemp       7802 non-null   float64
 7   humidity    7804 non-null   float64
 8   windspeed   7804 non-null   float64
 9   casual      7805 non-null   int64  
 10  registered  7805 non-null   int64  
 11  count       7805 non-null   int64  
dtypes: float64(5), int64(6), object(1)
memory usage: 731.8+ KB


В наличии 9 пропусков. Данные типов: float, int, str

### 7. Выведите стандартную описательную статистику по датафрему Сделайте выводы, касательно данных в столбцах: погода, влажность скорость ветра и количество (count). 

In [13]:
df.describe()

season      holiday   workingday      weather         temp  \
count  7805.000000  7805.000000  7805.000000  7803.000000  7803.000000   
mean      2.182319     0.027546     0.681871     1.419710    19.005967   
std       1.088684     0.163680     0.465780     0.642872     7.601849   
min       1.000000     0.000000     0.000000     1.000000     0.820000   
25%       1.000000     0.000000     0.000000     1.000000    13.120000   
50%       2.000000     0.000000     1.000000     1.000000    18.860000   
75%       3.000000     0.000000     1.000000     2.000000    25.420000   
max       4.000000     1.000000     1.000000     4.000000    38.540000   

             atemp     humidity    windspeed       casual   registered  \
count  7802.000000  7804.000000  7804.000000  7805.000000  7805.000000   
mean     22.377205    61.239493    13.276312    31.336195   131.398206   
std       8.330857    19.888351     8.449740    46.273109   127.076847   
min       0.760000     0.000000     0.000000     0.000000     0.000000   
25%      15.150000    46.000000     7.001500     3.000000    30.000000   
50%      22.725000    61.000000    12.998000    13.000000   100.000000   
75%      29.545000    77.000000    19.001200    40.000000   189.000000   
max      43.940000   100.000000    56.996900   367.000000   770.000000   

             count  
count  7805.000000  
mean    162.734401  
std     155.419640  
min       1.000000  
25%      35.000000  
50%     121.000000  
75%     238.000000  
max     873.000000

Преобладала ясная погода. 
Cредняя влажность 61%, но с разбросом примерно в 20 процентов.
Средняя скорость ветра 13.3 км/ч.
Максимальное кол-во велосипедов, арендованных за день = 873.

### 8. Изменяя параметры статистики, получите новые описательные результаты, и сделайте не менее трех утверждений на их основе. 

In [14]:
df.describe(percentiles=[0.15, 0.60, 0.85])

season      holiday   workingday      weather         temp  \
count  7805.000000  7805.000000  7805.000000  7803.000000  7803.000000   
mean      2.182319     0.027546     0.681871     1.419710    19.005967   
std       1.088684     0.163680     0.465780     0.642872     7.601849   
min       1.000000     0.000000     0.000000     1.000000     0.820000   
15%       1.000000     0.000000     0.000000     1.000000     9.840000   
50%       2.000000     0.000000     1.000000     1.000000    18.860000   
60%       2.000000     0.000000     1.000000     1.000000    21.320000   
85%       4.000000     0.000000     1.000000     2.000000    27.880000   
max       4.000000     1.000000     1.000000     4.000000    38.540000   

             atemp     humidity    windspeed       casual   registered  \
count  7802.000000  7804.000000  7804.000000  7805.000000  7805.000000   
mean     22.377205    61.239493    13.276312    31.336195   131.398206   
std       8.330857    19.888351     8.449740    46.273109   127.076847   
min       0.760000     0.000000     0.000000     0.000000     0.000000   
15%      12.880000    40.000000     6.003200     1.000000    11.000000   
50%      22.725000    61.000000    12.998000    13.000000   100.000000   
60%      25.000000    67.000000    15.001300    21.000000   130.000000   
85%      31.820000    84.000000    22.002800    62.000000   247.000000   
max      43.940000   100.000000    56.996900   367.000000   770.000000   

             count  
count  7805.000000  
mean    162.734401  
std     155.419640  
min       1.000000  
15%      14.000000  
50%     121.000000  
60%     163.000000  
85%     325.000000  
max     873.000000

В большинстве дней аренда велосипеов не превышала 247 штук, что сильно отличается от максимального значения. 
Большинство дней была ясная погода. 
Большинство дней скорость ветра не поднималась выше 15 км/ч. 

### 9. Выведите значения столбца «скорость ветра» начиная с 5 по 20. 

In [15]:
df["windspeed"][5:21]

5      6.0032
6      0.0000
7      0.0000
8      0.0000
9      0.0000
10    16.9979
11    19.0012
12    19.0012
13    19.9995
14    19.0012
15    19.9995
16    19.9995
17    19.0012
18    16.9979
19    16.9979
20    16.9979
Name: windspeed, dtype: float64

### 10. Выведите все значения столбца «temp», значения которых выше среднего. 

In [16]:
for i in df['temp']:
    if i >= df['temp'].mean():
        print(i)

21.32
22.96
23.78
24.6
22.96
21.32
19.68
22.14
24.6
24.6
22.96
23.78
22.14
19.68
19.68
21.32
20.5
20.5
24.6
27.06
27.06
27.06
27.06
26.24
25.42
24.6
23.78
22.14
21.32
19.68
20.5
21.32
22.14
22.14
20.5
19.68
19.68
19.68
19.68
19.68
20.5
21.32
20.5
21.32
20.5
20.5
21.32
22.14
22.96
26.24
27.06
27.88
27.88
28.7
27.88
27.06
25.42
25.42
25.42
24.6
24.6
23.78
22.96
22.14
21.32
21.32
20.5
20.5
20.5
20.5
19.68
20.5
22.14
24.6
26.24
27.88
30.34
31.16
31.16
30.34
29.52
28.7
28.7
28.7
27.88
26.24
25.42
25.42
22.14
22.14
20.5
19.68
19.68
20.5
21.32
22.14
21.32
21.32
20.5
21.32
22.96
24.6
24.6
21.32
19.68
20.5
20.5
21.32
21.32
20.5
20.5
19.68
20.5
21.32
22.96
22.96
24.6
24.6
26.24
29.52
30.34
30.34
30.34
29.52
29.52
27.88
27.06
26.24
23.78
25.42
25.42
24.6
23.78
22.96
22.14
22.14
22.14
19.68
20.5
21.32
22.96
22.14
20.5
19.68
19.68
20.5
21.32
22.14
22.96
22.96
24.6
24.6
24.6
22.96
22.14
20.5
19.68
20.5
20.5
21.32
21.32
22.14
22.14
22.14
21.32
20.5
21.32
21.32
20.5
20.5
20.5
21.32
20.5
21.32
22.14
22

### 11. Создайте еще один столбец в датафреме с именем «good_temp» и запишите в него результат сравнения текущей температуры со средним значением (True or False). 

In [20]:
def is_good(s) -> bool:
    return s > mean_t

mean_t = df['temp'].mean()
df['good_temp'] = df['temp'].map(is_good)
df

datetime  season  holiday  workingday  weather   temp  \
0     2011-01-01 00:00:00       1        0           0      1.0   9.84   
1     2011-01-01 01:00:00       1        0           0      1.0   9.02   
2     2011-01-01 02:00:00       1        0           0      1.0   9.02   
3     2011-01-01 03:00:00       1        0           0      1.0   9.84   
4     2011-01-01 04:00:00       1        0           0      1.0   9.84   
...                   ...     ...      ...         ...      ...    ...   
7800  2012-06-05 19:00:00       2        0           1      1.0  23.78   
7801  2012-06-05 20:00:00       2        0           1      1.0  22.96   
7802  2012-06-05 21:00:00       2        0           1      1.0  22.14   
7803  2012-06-05 22:00:00       2        0           1      1.0  22.96   
7804  2012-06-05 23:00:00       2        0           1      1.0  22.14   

       atemp  humidity  windspeed  casual  registered  count  good_temp  
0     14.395      81.0     0.0000       3          13     16      False  
1     13.635      80.0     0.0000       8          32     40      False  
2     13.635      80.0     0.0000       5          27     32      False  
3     14.395      75.0     0.0000       3          10     13      False  
4     14.395      75.0     0.0000       0           1      1      False  
...      ...       ...        ...     ...         ...    ...        ...  
7800  27.275      46.0    16.9979      68         445    513       True  
7801  26.515      49.0    15.0013      44         371    415       True  
7802  25.760      56.0    11.0014      30         253    283       True  
7803  26.515      52.0     7.0015      22         171    193       True  
7804  25.760      56.0     6.0032      20          93    113       True  

[7805 rows x 13 columns]

### 12. Выведите те строки, в которых есть пустые значения (в любом из столбцов). 

In [21]:
df[df.isnull().any(axis=1)]

datetime  season  holiday  workingday  weather   temp  \
273   2011-01-12 20:00:00       1        0           1      NaN   6.56   
281   2011-01-13 04:00:00       1        0           1      1.0   5.74   
343   2011-01-15 19:00:00       1        0           0      2.0  13.12   
727   2011-02-14 02:00:00       1        0           1      1.0  14.76   
977   2011-03-05 14:00:00       1        0           0      NaN  19.68   
1049  2011-03-08 16:00:00       1        0           1      1.0  15.58   
1114  2011-03-11 12:00:00       1        0           1      2.0    NaN   
1115  2011-03-11 13:00:00       1        0           1      1.0    NaN   
1179  2011-03-14 07:00:00       1        0           1      2.0  11.48   

       atemp  humidity  windspeed  casual  registered  count  good_temp  
273    6.820      50.0    22.0028       2          54     56      False  
281    6.060      50.0        NaN       0           4      4      False  
343      NaN      39.0    16.9979      14          60     74      False  
727   16.665       NaN    19.9995       0           2      2      False  
977   23.485      59.0    19.0012      80         118    198       True  
1049     NaN      27.0    11.0014      26          82    108      False  
1114  15.910      53.0    16.9979      15          96    111      False  
1115  16.665      53.0    16.9979      15          92    107      False  
1179     NaN      61.0     0.0000       4          84     88      False

### 13. Выведите первые 5 строк, для которых выполняется составное условие: температура > 20 и скорость ветра > 20. 

In [29]:
new_df = df[(df['temp'] > 20) & (df['windspeed'] > 20)]
new_df.head(5)

datetime  season  holiday  workingday  weather   temp   atemp  \
737  2011-02-14 12:00:00       1        0           1      1.0  22.96  26.515   
738  2011-02-14 13:00:00       1        0           1      1.0  23.78  27.275   
739  2011-02-14 14:00:00       1        0           1      1.0  24.60  29.545   
740  2011-02-14 15:00:00       1        0           1      1.0  22.96  26.515   
741  2011-02-14 16:00:00       1        0           1      1.0  21.32  25.000   

     humidity  windspeed  casual  registered  count  good_temp  
737      22.0    30.0026      27          99    126       True  
738      19.0    26.0027      27          93    120       True  
739      15.0    32.9975      14          76     90       True  
740      21.0    43.9989      19          71     90       True  
741      27.0    31.0009      16         102    118       True

### 14. Выведите значения всех полей в третьей строке списка полученного на предыдущем шаге

In [30]:
new_df.iloc[2]

datetime      2011-02-14 14:00:00
season                          1
holiday                         0
workingday                      1
weather                       1.0
temp                         24.6
atemp                      29.545
humidity                     15.0
windspeed                 32.9975
casual                         14
registered                     76
count                          90
good_temp                    True
Name: 739, dtype: object

### 15. Выведите датафрейм, отсортированный по полю count, в обратном порядке. 

In [31]:
df.sort_values(by='count', ascending=False)

datetime  season  holiday  workingday  weather   temp  \
7606  2012-05-16 17:00:00       2        0           1      1.0  29.52   
7631  2012-05-17 18:00:00       2        0           1      1.0  27.06   
7630  2012-05-17 17:00:00       2        0           1      1.0  27.06   
7798  2012-06-05 17:00:00       2        0           1      1.0  24.60   
7607  2012-05-16 18:00:00       2        0           1      1.0  29.52   
...                   ...     ...      ...         ...      ...    ...   
942   2011-03-04 03:00:00       1        0           1      2.0   7.38   
919   2011-03-03 04:00:00       1        0           1      1.0   6.56   
918   2011-03-03 03:00:00       1        0           1      1.0   8.20   
916   2011-03-03 01:00:00       1        0           1      1.0   9.84   
6516  2012-03-09 04:00:00       1        0           1      3.0  16.40   

       atemp  humidity  windspeed  casual  registered  count  good_temp  
7606  33.335      51.0    12.9980     104         769    873       True  
7631  31.060      27.0    11.0014     134         734    868       True  
7630  31.060      31.0     8.9981     133         719    852       True  
7798  31.060      43.0     0.0000      86         764    850       True  
7607  33.335      51.0    15.0013      97         749    846       True  
...      ...       ...        ...     ...         ...    ...        ...  
942    8.335      74.0    16.9979       0           1      1      False  
919    6.060      31.0    32.9975       0           1      1      False  
918    8.335      27.0    31.0009       0           1      1      False  
916    9.850      30.0    31.0009       0           1      1      False  
6516  20.455      66.0    19.0012       0           1      1      False  

[7805 rows x 13 columns]

### 16. Создайте еще один датафрем на основе описания поля season (1 = spring, 2 = summer, 3 = fall, 4 = winter). 

In [34]:
df_seasons = pd.DataFrame([[1, 'spring'], [2, 'summer'], [3, 'fall'], [4, 'winter']], columns=['season', 'name_season'])
df_seasons

season name_season
0       1      spring
1       2      summer
2       3        fall
3       4      winter

### 17. Соедините исходный датафрем с даатфремом, созданным на предыдущем шаге по указанному выше правилу. Выведите полученный результат. 

In [37]:
new_df = pd.merge(df, df_seasons)
new_df

datetime  season  holiday  workingday  weather   temp  \
0     2011-01-01 00:00:00       1        0           0      1.0   9.84   
1     2011-01-01 01:00:00       1        0           0      1.0   9.02   
2     2011-01-01 02:00:00       1        0           0      1.0   9.02   
3     2011-01-01 03:00:00       1        0           0      1.0   9.84   
4     2011-01-01 04:00:00       1        0           0      1.0   9.84   
...                   ...     ...      ...         ...      ...    ...   
7800  2012-06-05 19:00:00       2        0           1      1.0  23.78   
7801  2012-06-05 20:00:00       2        0           1      1.0  22.96   
7802  2012-06-05 21:00:00       2        0           1      1.0  22.14   
7803  2012-06-05 22:00:00       2        0           1      1.0  22.96   
7804  2012-06-05 23:00:00       2        0           1      1.0  22.14   

       atemp  humidity  windspeed  casual  registered  count  good_temp  \
0     14.395      81.0     0.0000       3          13     16      False   
1     13.635      80.0     0.0000       8          32     40      False   
2     13.635      80.0     0.0000       5          27     32      False   
3     14.395      75.0     0.0000       3          10     13      False   
4     14.395      75.0     0.0000       0           1      1      False   
...      ...       ...        ...     ...         ...    ...        ...   
7800  27.275      46.0    16.9979      68         445    513       True   
7801  26.515      49.0    15.0013      44         371    415       True   
7802  25.760      56.0    11.0014      30         253    283       True   
7803  26.515      52.0     7.0015      22         171    193       True   
7804  25.760      56.0     6.0032      20          93    113       True   

     name_season  
0         spring  
1         spring  
2         spring  
3         spring  
4         spring  
...          ...  
7800      summer  
7801      summer  
7802      summer  
7803      summer  
7804      summer  

[7805 rows x 14 columns]

### 18. Сделайте группировку по временам года и количеству (count). 

In [38]:
new_df[['name_season', 'count']].groupby('name_season').count()

count
name_season       
fall          1365
spring        2676
summer        2397
winter        1367

### 19. Создайте pivot_table по любым столбцам и опишите полученный результат. 

Средняя температура в рабочие и выходные дни в зависимости от времени года.

In [46]:
pivot_t = new_df.pivot_table('temp', 'workingday' , 'name_season', aggfunc='mean')
pivot_t

name_season       fall     spring     summer     winter
workingday                                             
0            27.912269  12.371304  22.156532  15.464912
1            28.429303  12.621163  22.380593  17.182195

### 20. Получите процентное соотношение count в зависимости от времен года. 

In [48]:
pivot_t = new_df.pivot_table('count', 'name_season', aggfunc='sum')
pivot_t.div(pivot_t.sum(0))*100


count
name_season           
fall         20.096021
spring       24.571111
summer       38.673786
winter       16.659082